lib/hook.mli

```diff
@@ -27,3 +32,6 @@ val get : 'a value -> 'a
 
 val set : 'a t -> 'a -> unit
 (** Register a hook. Assertion failure if already registered. *)
+
+val set_default : 'a t -> 'a -> unit
```

Comment:
This is kinda strange IMO. Wouldn't it make more sense to expose the TC resolver as a regular ref?

Reply:
I don't think we should expose this as a hook either. Ideally, it would even be better to have a libobject registering the currently used TC algorithm, otherwise that means we lose modularity of development. What if you want to use two libraries using different hooks?

Reply:
I agree that hooks are problematic, see for example discussion on https://github.com/coq/coq/pull/8768

However I feel that this PR is OK for now, as it doesn't really change the current situation.

Reply:
My motivation was to enable an experiment, not to streamline this. A bit like unicoq does.

In particular if two libraries use different TC engines I don't see how a libobject would solve the problem. Sure, it would help to select one or the other, but this is just the tip of the iceberg. When you combine two lemmas, they may need the power of both engine... so neither would do.

Reply:
Indeed I share @gares view ; @ppedrot , @SkySkimmer are you OK with letting this in for experimentation purposes for the moment then?

Reply:
Make it a ref as I said
Hook is basically a workaround for not having recursive modules across files
If you want to be able to override something ref is the right type

Reply:
What does that mean?

Reply:
That sounds pretty overengineered, just use ref

Reply:
I don't want random refs in the code, they end up being abuse to hide design problems, and that's something that introduces bugs in the upper layers easily as they need to rely on referentially transparent code.

Reply:
>> I mean having a module such as SettableRef.t that is used to set such refs.
>
> That sounds pretty overengineered, just use ref

I'm not against a ref per se, but Hooks are already there and this change is trivial. IMO either we purge hooks (again, not against, but not the job of this PR), or we stay with Hooks and this PR just refines their APIs.

Reply:
What about if we keep the hook and add a comment that it should go into libobject in the future?

Reply:
> What about if we keep the hook and add a comment that it should go into libobject in the future?

That would be misleading as as said above libobject won't be able to help with this

Reply:
Using Hook is useful to mean "this thing s not overrideable".
That is why this PR should just switch to ref.

Reply:
> Using Hook is useful to mean "this thing s not overrideable".

Well, this is not how hooks work today. Maybe it was an oversight, but:
- If the hook has a default value, then this value is overridable *once*.
- if the hook has no default value, then it is settable *once* (and if you don't set it, boom).

This PR restores a bit of symmetry:
- a hook can be given a default value (immediately or later).
- the default value (be it set or not) can be overridden *once*.

If hooks were not overrideable then unicoq would not work, see the Hook for evarconv.

I don't have a strong opinion on the best solution. @SkySkimmer if you do have, please do a request for changes.

Reply:
I agree with @SkySkimmer analysis: a hook is a workaround for (lack of) recursive modules. Don't use a hook for things that are by design settable.

Reply:
>If hooks were not overrideable then unicoq would not work, see the Hook for evarconv.

That should be a ref too, same for the ones in stm.ml.

There is 1 more default value user in evd for is_maybe_typeclass, that one is set in typeclasses.ml so the default is actually never used.

Reply:
OK, so we do agree I remove the `Default` case in the type. Since it is either unset or set.